ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,382 @@
+          )
+      )
+
+(** - Direct: request doesn't pass through the cache
+    - Engaged: both side coordinate through the queue, writer ends the mode
+      when the queue has been filled.
+    - Disengaged: writer ignores the queue, reader empties it and the cache;
+      then it changes the mode to engaged.
+*)
+type state = Direct | Engaged | Disengaged
+
+type channel = {
+    queue: t Lwt_bounded_stream.t
+  ; push: t option -> unit option
+  ; lock: Lwt_mutex.t (* lock for the states *)
+  ; mutable state: state
+}
+
+(*
+   Notes:
+   - uses Mtime.t to force usage of monotonic time
+   - This means that between runs (and reboots) cached stated is lost if not
+     persisted first.
+   IDEA: carryover: read contents of cache and "convert it" to the current run
+
+   TODO:
+     - Add startup step to convert existing content to new time
+     - Exponential backoff on xapi push error
+     - Limit error logging on xapi push error: once per downtime is enough
+ *)
+
+module Writer : sig
+  val with_cache :
+       direct:(t -> string -> (unit, exn) Lwt_result.t)
+    -> Types.Service.t
+    -> channel
+    -> ((t -> string -> unit Lwt.t) -> 'a Lwt.t)
+    -> 'a Lwt.t
+  (** [with_cache ~direct typ context] creates a cache for content of type [typ]. The
+      cache is writable through the function [context], which is provided a
+      writing function to persist to the cache.
+
+      Example:
+        Xapi_guard.Disk_cache.(Writer.with_cache Tpm (upload session_cache)
+        @@ fun write_tpm -> write_tpm (uuid, time, key) contents
+    *)
+end = struct
+  let mkdir_p ?(perm = 0o755) path =
+    let rec loop acc path =
+      let create_dir () = Lwt_unix.mkdir path perm in
+      let create_subdirs () = Lwt_list.iter_s (fun (_, f) -> f ()) acc in
+      Lwt.try_bind create_dir create_subdirs (function
+        | Unix.(Unix_error (EEXIST, _, _)) ->
+            (* create directories, parents first *)
+            create_subdirs ()
+        | Unix.(Unix_error (ENOENT, _, _)) ->
+            let parent = Filename.dirname path in
+            loop ((path, create_dir) :: acc) parent
+        | exn ->
+            let msg =
+              Printf.sprintf {|Could not create directory "%s" because: %s|}
+                path (Printexc.to_string exn)
+            in
+            Lwt.fail (Failure msg)
+        )
+    in
+    loop [] path
+
+  let files_in_existing dir =
+    let create_dir = function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          let* () = mkdir_p dir ~perm:0o700 in
+          Lwt.return []
+      | e ->
+          raise e
+    in
+    files_in dir ~otherwise:create_dir
+
+  let persist_to ~filename ~contents =
+    let flags = Unix.[O_WRONLY; O_CREAT; O_SYNC] in
+    let perm = 0o600 in
+    let persist out_chan = Lwt_io.write out_chan contents in
+    Lwt_io.with_file ~flags ~perm ~mode:Output filename persist
+
+  let write_contents ~direct root queue (uuid, now, key) contents =
```

Comment:
My type inference seems to be unable to infer that `direct` has type `t -> string -> ...` but instead gives `Uuidm.t * Mtime.t * Types.Tpm.key -> ...`. Maybe use cache as the argument, and then do `let (uuid, now, key) = cache` to satisfy the type inference?

Reply:
It's not a failure, the type inference gives me 
```ocaml
direct:
  (Uuidm.t * Mtime.t * Types.Tpm.key ->
  string ->
  (unit, exn) result Lwt.t) ->
string ->
channel ->
Uuidm.t * Mtime.t * Types.Tpm.key ->
string ->
unit Lwt.t
```
but ideally we want 
```ocaml 
direct:
  (t ->
  string ->
  (unit, exn) result Lwt.t) ->
string ->
channel ->
t ->
string ->
unit Lwt.t
```

Reply:
maybe it's to do with my lsp? but anyway it's a minor thing so you can leave it as it is

Reply:
The types are equivalent, so I don't think it's an issue. For example it can display `UUIDSet.elt` instead of `Uuidm.t` in some places. Or `dbginfo` instead of `string`